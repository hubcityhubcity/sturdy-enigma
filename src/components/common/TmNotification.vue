<template>
  <div v-if="active" :class="cssClass" @click="deactivate()">
    <header>
      <div v-if="icon" class="icon">
        <i class="material-icons notranslate">{{ icon }}</i>
      </div>
      <div v-if="title" class="title">
        {{ title }}
      </div>
      <menu>
        <div v-if="time" class="time">
          {{ fromNow }}
        </div>
        <i class="close material-icons">close</i>
      </menu>
    </header>
    <div class="body">
      {{ body }}
    </div>
  </div>
</template>

<script>
import moment from "moment"
export default {
  name: `TmNotification`,
  props: {
    type: {
      type: String,
      default: null,
    },
    layout: {
      type: String,
      default: null,
    },
    icon: {
      type: String,
      default: null,
    },
    time: {
      type: Number,
      required: true,
    },
    title: {
      type: String,
      default: null,
    },
    body: {
      type: String,
      required: true,
    },
  },
  data: () => ({
    duration: 5000,
    active: true,
  }),
  computed: {
    fromNow() {
      return moment(this.time).fromNow()
    },
    cssClass() {
      let value = `tm-notification`
      if (this.type) value += ` tm-notification-${this.type}`
      return value
    },
  },
  mounted() {
    this.setDeactivation()
  },
  methods: {
    deactivate() {
      this.active = false
    },
    setDeactivation() {
      if (!this.layout || this.layout === `banner`) {
        // notification active duration is 5 seconds - (time since creation)
        const activeDuration = this.duration - (Date.now() - this.time)

        // disable visibility if it's an old notification
        if (activeDuration < 0) {
          this.active = false
          return
        }

        // otherwise self destruct after duration
        setTimeout(() => this.deactivate(), activeDuration)
      }
    },
  },
}
</script>

<style>
.tm-notification {
  background: var(--app-nav);
  font-size: 0.75rem;
  cursor: pointer;
  user-select: none;
  margin: 0.5rem 0.5rem 0;
  border-radius: 0.25rem;
  border: 1px solid var(--bc-dim);
}

.tm-notification header {
  display: flex;
  align-items: center;
  padding: 0 0.375rem;
  height: 2em;
  border-radius: 0.25rem 0.25rem 0 0;
  background: var(--success);
}

.tm-notification.tm-notification-warning header {
  background: var(--warning);
}

.tm-notification.tm-notification-error header {
  background: var(--danger);
}

.tm-notification header .icon,
.tm-notification header .title {
  color: var(--bright);
}

.tm-notification header .icon {
  width: 1rem;
  display: flex;
  align-items: center;
  margin-right: 0.5rem;
}

.tm-notification header .title {
  flex: 1;
  font-weight: 500;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  padding-right: 0.375rem;
}

.tm-notification header menu {
  color: rgba(255, 255, 255, 0.67);
  font-size: 0.75rem;
}

.tm-notification header menu .close {
  display: none;
}

.tm-notification .body {
  padding: 0.375rem;
  color: var(--txt);
}

.tm-notification:hover menu .time {
  display: none;
}

.tm-notification:hover menu .close {
  display: block;
  cursor: pointer;
}

@media screen and (min-width: 360px) {
  .tm-notification {
    font-size: 0.875rem;
    margin: 0.625rem 0.625rem 0;
  }

  .tm-notification header {
    padding: 0 0.5rem;
  }

  .tm-notification .body {
    padding: 0 0.5rem;
  }
}

@media screen and (min-width: 400px) {
  .tm-notification {
    margin: 0.75rem 0.75rem 0;
  }

  .tm-notification header {
    padding: 0 0.75rem;
  }

  .tm-notification .body {
    padding: 0.75rem;
  }
}

@media screen and (min-width: 720px) {
  .tm-notification {
    margin: 1rem 1rem 0;
  }
}
</style>
