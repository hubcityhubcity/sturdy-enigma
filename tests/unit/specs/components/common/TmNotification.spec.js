import { mount } from "@vue/test-utils"
import TmNotification from "common/TmNotification"

describe(`TmNotification.vue`, () => {
  const notifications = [
    {
      time: Date.now() - 1,
      body: `asdf`,
    },
    {
      title: `title`,
      time: Date.now() - 1,
      body: `sadf`,
    },
    {
      icon: `search`,
      time: Date.now() - 1,
      body: `sadf`,
    },
    {
      type: `warning`,
      time: Date.now() - 1,
      body: `sadf`,
    },
    {
      type: `error`,
      time: Date.now() - 1,
      body: `sadf`,
    },
    {
      layout: `alert`,
      time: Date.now() - 1,
      body: `sadf`,
    },
  ]

  it(`has the expected html structure`, () => {
    notifications.forEach((notification) => {
      const wrapper = mount(TmNotification, {
        propsData: { ...notification },
      })
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  it(`closes after a certain timeout automatically`, () => {
    jest.useFakeTimers()
    const wrapper = mount(TmNotification, {
      propsData: {
        time: 44000,
        body: `asdf`,
      },
    })
    expect(wrapper.vm.active).toBe(true)
    jest.runAllTimers()
    expect(wrapper.vm.active).toBe(false)
  })

  it(`is hidden if timed out`, () => {
    const wrapper = mount(TmNotification, {
      propsData: {
        time: 100,
        body: `asdf`,
      },
    })
    expect(wrapper.vm.active).toBe(false)
  })
})
