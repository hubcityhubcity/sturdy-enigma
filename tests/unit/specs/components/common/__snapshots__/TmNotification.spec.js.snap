// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmNotification.vue has the expected html structure 1`] = `
<div
  class="tm-notification"
>
  <header>
    <!---->
     
    <!---->
     
    <menu>
      <div
        class="time"
      >
        
        a few seconds ago
      
      </div>
       
      <i
        class="close material-icons"
      >
        close
      </i>
    </menu>
  </header>
   
  <div
    class="body"
  >
    
    asdf
  
  </div>
</div>
`;

exports[`TmNotification.vue has the expected html structure 2`] = `
<div
  class="tm-notification"
>
  <header>
    <!---->
     
    <div
      class="title"
    >
      
      title
    
    </div>
     
    <menu>
      <div
        class="time"
      >
        
        a few seconds ago
      
      </div>
       
      <i
        class="close material-icons"
      >
        close
      </i>
    </menu>
  </header>
   
  <div
    class="body"
  >
    
    sadf
  
  </div>
</div>
`;

exports[`TmNotification.vue has the expected html structure 3`] = `
<div
  class="tm-notification"
>
  <header>
    <div
      class="icon"
    >
      <i
        class="material-icons notranslate"
      >
        search
      </i>
    </div>
     
    <!---->
     
    <menu>
      <div
        class="time"
      >
        
        a few seconds ago
      
      </div>
       
      <i
        class="close material-icons"
      >
        close
      </i>
    </menu>
  </header>
   
  <div
    class="body"
  >
    
    sadf
  
  </div>
</div>
`;

exports[`TmNotification.vue has the expected html structure 4`] = `
<div
  class="tm-notification tm-notification-warning"
>
  <header>
    <!---->
     
    <!---->
     
    <menu>
      <div
        class="time"
      >
        
        a few seconds ago
      
      </div>
       
      <i
        class="close material-icons"
      >
        close
      </i>
    </menu>
  </header>
   
  <div
    class="body"
  >
    
    sadf
  
  </div>
</div>
`;

exports[`TmNotification.vue has the expected html structure 5`] = `
<div
  class="tm-notification tm-notification-error"
>
  <header>
    <!---->
     
    <!---->
     
    <menu>
      <div
        class="time"
      >
        
        a few seconds ago
      
      </div>
       
      <i
        class="close material-icons"
      >
        close
      </i>
    </menu>
  </header>
   
  <div
    class="body"
  >
    
    sadf
  
  </div>
</div>
`;

exports[`TmNotification.vue has the expected html structure 6`] = `
<div
  class="tm-notification"
>
  <header>
    <!---->
     
    <!---->
     
    <menu>
      <div
        class="time"
      >
        
        a few seconds ago
      
      </div>
       
      <i
        class="close material-icons"
      >
        close
      </i>
    </menu>
  </header>
   
  <div
    class="body"
  >
    
    sadf
  
  </div>
</div>
`;
